tests/integration/nm/ipv4_test.py

```diff
@@ -0,0 +1,82 @@
+#
+# Copyright 2019 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+from libnmstate import nm
+
+from .testlib import mainloop
+
+
+ETH1 = 'eth1'
+
+IPV4_ADDRESS1 = '192.0.2.251'
+
+
+def test_interface_ipv4_change(eth1_up):
```

Comment:
IMHO a test for reapply should be in its own test module. It is not really clear taht the IPv4 test module would test reapply. Also the test does not really verify that reapply was used. Since there is a fallback mechanism to use normal activation instead, it would be great to actually verify that reapply works as expected.

Reply:
Your feedback is related to the "strict" reapply I mentioned in my previous response.
Without a "strict" version, I do not know how to check that reapply worked or not.

I also do not see a great benefit in testing the reapply separately, as I would just have to use a subject like ipvX, wired and so on... When one of those is checked, it implicitly checks activate or reapply.

In the context of nmstate, we will always use reapply (see the following changes/PRs) with the fallback.

Reply:
IMHO it makes sense to explicitly test all supported changes with reapply. In the current test setup, the actual test case (`test_interface_ipv4_change()`) does not contain any indication that reapply is required and also does not verify it. AFAIU, only the line `nm.device.reapply(nmdev, conn.profile)` might make the test use reapply. However, this line could just be changed to issue a normal activation or the reapply method could be accidentally changed to never really do a reapply (for example because the check whether reapply would work accidentally always returns True) and the test case would not fail. So the test case only checks that using `nm.device.reapply()` does not break adding an IPv4 address but it tests nothing about the actual functionality.

Also IMHO a strict method is not the way to check it but some assertion that for example the link did not go down for the interface (which is AFAIU why we want reapply).

Reply:
Btw. having a user story like `I can change the IP address for an interface without the link going down.` and developing a test that shows that this user story is supported might help here.

Reply:
The vdsm test is not part of Nmstate. Therefore the current test does not show that the code actually delivers something new (reapply). Even if you split it into three functions, the tests will not show that the functions actually use reapply. This can only be guessed by looking at the actual implementation.

Regarding doubling the tests, if we want to have meaningful tests that check both the normal activation (which would be needed when the profile is created for the first time I guess) and reapply (which would be changing a profile), we need twice as much tests as if we only test one of these scenarios. It might also be ok to do this in one test, however there should be assertions that ensure that when changing the profile, traffic was not disrupted.

But in the end, since this is an integration test, whether or not reapply is used should not matter. What matters is the actual behavior which would be no traffic disruption when changing the IP address (according the the commit messages). If you do not want to do the test, I will not force you to do it. Nevertheless, it is a flaw in the current test and does not seem to hold up to the standards that you seem to have.

Reply:
I agree with two points here.
The first, you explicitly gave a scenario where activation and not reapply is needed: create a new interface.
The second, we need to check using monitor, but that is too large imo to implement now.